### PR TITLE
Protect against sorting with a bin-edge coordinate

### LINKED ISSUE #3086: `sc.sort` has strange behaviour with bin edge coordinates
Sorting a data array with a midpoints coord works as expected.
However, making the same data array with a bin edge coord has a strange behaviour.
```Py
a = sc.DataArray(data=sc.array(dims=['x'], values=np.random.random(20)),
                 coords={'x': sc.linspace('x', 80, 33, 21, unit='m')})
sc.sort(a, key='x')
```
`sc.sort` drops the coord, probably because in principle there is no way to keep the bin edge coordinate if the edges are not monotonous.
However, if they are monotonous, the coord could be kept in the result.

In addition, the above returns a data array with data than now has shape 21 instead of 20, and the first value in the array is garbage (e.g. 1.0e-300).

### PR DESCRIPTION
**Pull request by nvaytet**
We throw an error in `sc.sort` when the key for sorting does not have the same shape as the input Variable to be sorted.

Fix #3086

### CODE REVIEW

File: lib/dataset/sort.cpp [hunk 1]
```diff
@@ -59,16 +59,25 @@ Variable indices_for_sorting(const Variable &key, const SortOrder order) {
       core::time_point>::apply<IndicesForSorting>(key.dtype(), key, order);
 }
 
+void require_same_shape(const Dimensions &a, const Dimensions &b,
+                        const Dim dim) {
+  if (a[dim] != b[dim])
+    throw except::DimensionError(
+        "Cannot sort based on key with different shape.");
```
Comment (SimonHeybrock) [inline lib/dataset/sort.cpp:66]: I think shape can be different (the sorted data may be 2d?). Refer to bin-edges instead?

Comment (nvaytet) [reply lib/dataset/sort.cpp:66]: Ok about not using just 'shape', but it's not only bin edges. The following is also an issue:
```Py
sc.sort(sc.arange('x', 10.), key=sc.arange('x', 50.))
```
gives
```
<scipp.Variable> (x: 50)    float64  [dimensionless]  [0, 1, ..., 8.18433e-85, 4.56532e-85]
```

Comment (SimonHeybrock) [reply lib/dataset/sort.cpp]: Then say "the length of the sort key"?

Comment (nvaytet) [reply lib/dataset/sort.cpp]: See the latest commits?

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**